### PR TITLE
Insufficient material draw

### PR DESCRIPTION
**Pull request by stockfishdeveloper**
Detect positions drawn by insufficient material.

vs Master:
+756 -218 =26
Number of games: 1000
Winning fraction: 0.769
Elo difference: +208.926

I suggest you test my changes as well because they seem to be giving +200 ELO?!

### CODE REVIEW

File: src/position.cc [hunk 1]
```diff
@@ -595,3 +595,13 @@ void Position::print() const
...
+
+bool Position::insufficient_material() const
+{
+	if(bb::count(occ()) > 3)
+		return false;
+	if(bb::count(occ(KNIGHT)) == 1)
+		return true;
+	if(bb::count(occ(BISHOP)) == 1)
+		return true;
+}
```
Comment (lucasart) [inline src/position.cc]: does this compile without warning ?

**Comment by lucasart** [comment]
This 208 elo makes no sense, even with a small sample of 1000 games. Can you explain how you ran the test ? I strongly recommended you use cutechess-cli , and paste the command line.

Also, elo isn't the way to test this change, because it's worth zero elo (assuming elo measure is done correctly). You need to test the functionality, verify that it detects draws when it should, and doesn't detect draws when not expected.

**Comment by stockfishdeveloper** [comment]
Yep. I had the return statement in a previous attempt, I forgot it here. 
You're right, this should not be an ELO patch, it's strange I got those results.
I was testing it in LittleBlitzer with 5 sec + 50 ms increment. I'll try it in CuteChess-cli.

It does recognize drawn positions before they appear on the board, so it's functionally sound.
I tested the following positions among others:
8/2r1k3/8/3PK3/8/3B4/8/8 w - - 0 1
8/8/8/5B2/7k/5K1p/8/8 w - - 0 1
They both return a draw score.

I'll test it again and I noted that it might be good to put all of the draw detection in one wrapper function.

File: src/position.cc (hunk 1 above)
Comment (stockfishdeveloper) [inline src/position.cc]: No, I stupidly disabled warnings. Good catch though!

**Comment by lucasart** [comment]
Ok. That result makes more sense, and is indeed within the error bar.

But you're forgetting the KK case.

**Comment by stockfishdeveloper** [comment]
Well, any GUI that allows a KK position on the board before adjudication is bogus. In internal nodes I would expect the eval of KK to be literally 0. Therefore we shouldn't need a KK case?

If we need to put it in I can just add:

if(bb::count(occ()) == 2)

**Comment by lucasart** [comment]
Did you verify that KK eval is zero ?
Do you understand the difference between recognizing a draw in the search, and the eval returning zero (aside from the fact that DrawScore is not zero) ?

**Comment by stockfishdeveloper** [comment]
The first few depths are not 0, all the following ones give -10, our DrawScore.

Absolutely. The search can return 0 when a non-drawn position is on the board. Like for example, in human chess we say that a position is "equal", meaning the eval is 0. However, that doesn't mean that the position has no chances for either side. It **only** means that the chances for both sides are equal.

I'll add the KK case anyway--it can't hurt anything.

**Comment by lucasart** [comment]
If you want to learn, you need to be curious and try to understand. It's easy and tempting to hide problems under the carpet. Let's see:
* you noticed that the search result in KK is not zero at low depth. Do you understand why?
* you noticed that it eventually reaches the contempt score as we search deep enough. Do you understand why?
* returning zero in search or in eval is not the same. You need to understand how alpha beta and particularly qsearch work, first.

**Comment by stockfishdeveloper** [comment]
I thank you for your interest in educating me!

I'm not sure why we don't get the contempt score at lower depths--I'm interested in finding out.
Well, we get the contempt score as we go deeper because we're recognizing a drawn position and returning a draw score--I don't know why at lower depths we're not getting that score.
Right--if the eval returns 0, it doesn't mean the search returns 0. We just note that that child node equals 0 and depending on the values of alpha and beta we can prune.

So why does it take several plies to reach the contempt score?

**Comment by lucasart** [comment]
You're jumping to conclusions without following a scientific method. So let's use some method:
* what does the eval return in a KK position? Please don't confuse search and eval, I mean the actual eval with no search whatsoever.
* what is the search expected to return at depth <= 0, in a KK position ? (hint: there are no captures available).
* what happens at depth = 1 ?
* etc.

**Comment by stockfishdeveloper** [comment]
- K on e1 and K on e8 has an eval of 0. K on e4 and K on e8 has an eval of 49. There must be a bonus for king in the center.....

- The search is supposed to see depth <= 0 and go into quiescence search. Since there are no captures, promotions, or other drastic moves, the Qsearch probably searches no nodes and returns the eval of the position. I would expect it to return the eval of the position at depth <= 0.

- At depth = 1, all legal moves should be searched, the best one for whose turn it is gets returned. Since contempt is -10, all legal moves should result in a position where it's KK and the ply should be worth -10. But before we reach the part checking for insufficient material we go into Qsearch, right? So we don't know that position is drawn and head right into Qsearch which gives is a value of 49 for that node. This would explain why we get the wrong value returned from a 1-ply search, but what about the higher depths?

- At depth 2, for example, we don't go into Qsearch so we do know the position is drawn. Now why don't we get a score of -10 returned from the search?

**Comment by lucasart** [comment]
* Yes, there is a bonus for having a King in the center in the endgame. And a malus for having a King in the center in the opening. And linearly interpolated in between opening and endgame, based on remaining material. It's call PSQT (piece on square table), and is the first part of any chess engine eval, after material.
* Yes, search returns eval at depth <= 0, because there are no captures. It's the "Qsearch stand pat" part (see search.cc).
* At depth=1, the search has 1 move to improve the eval of the side to move. So it returns whichever king move gives a beter eval. Typically a step towards the center, because the eval of KK simplifies to PSQT only (other eval components are zero). So it should be slightly higher than depth=0 which returns eval.
* depth=2, same twice, for the side to move and the opponent. IOW: I make a move towards the center, but so does my opponent, and now we look at the eval. Should be similar to depth=0.
* etc. You see a depth parity effect. This is somewhat compensated by the Tempo in the search by the way.
* eventually, the search understands that the side with the worst eval (furthest from the center) can force a repetition or threaten to penetrate the center and equalise the eval. So the score should converge towards the `[-10,0]` interval eventually.

File: src/position.cc [hunk 2]
```diff
@@ -595,3 +595,16 @@ void Position::print() const
...
+	if(bb::count(occ()) > 3)
+		return false;
+	if(bb::count(occ(KNIGHT)) == 1)
+		return true;
+	if(bb::count(occ(BISHOP)) == 1)
+		return true;
+	if(bb::count(occ()) == 2)
+		return true;
+	return false;
+}
```
Comment (lucasart) [inline src/position.cc]: Can you write this in a cleaner way ?

One line should suffice, and you don't need any `if` and `count()`.

**Comment by stockfishdeveloper** [comment]
So basically you're saying that this is normal and we're actually detecting a draw the right way?

I'll simplify the PR in a minute.

File: src/position.cc (hunk 2 above)
Comment (stockfishdeveloper) [inline src/position.cc]: OK, I cut down a few lines. It's functional integrity was retained.
I could only get it shorter if I added a function to the Position class like Minors() to return the bishop and knight occupancy.
I think it's fine the way it is though.

File: src/position.cc [hunk 3]
```diff
@@ -595,3 +595,12 @@ void Position::print() const
         std::cout << std::endl;
     }
 }
+
+bool Position::insufficient_material() const
+{
+	if(bb::count(occ()) <= 3) {
+		if(bb::count(occ(KNIGHT)) == 1 || bb::count(occ(BISHOP)) == 1)
+		return true;
+	}
```
Comment (lucasart) [inline src/position.cc]: you've broken KK again.

**Comment by stockfishdeveloper** [comment]
Ok, fixed and I'm closing this PR and open another to make it easier to merge.

File: src/position.cc (hunk 3 above)
Comment (stockfishdeveloper) [inline src/position.cc]: Ok, fixed. Ready to merge?

**Comment by lucasart** [comment]
Sorry, but your code smells. It may be shorter by a couple of lines, but it's not simpler. I'll write it to show you what I mean.

**Comment by stockfishdeveloper** [comment]
OK.
I don't know this codebase as well as my own engine.
I still have a lot to learn--I'd like to see how you could make it simpler.
